snap/validate.go

```diff
@@ -494,9 +518,18 @@ func ValidateLayout(layout *Layout) error {
 		if err := ValidatePathVariables(oldname); err != nil {
 			return fmt.Errorf("layout %q uses invalid symlink old name %q: %s", layout.Path, oldname, err)
 		}
+		oldname = si.ExpandSnapVariables(oldname)
 		if !isAbsAndClean(oldname) {
 			return fmt.Errorf("layout %q uses invalid symlink old name %q: must be absolute and clean", layout.Path, oldname)
 		}
+		// Symlinks *must* use $SNAP, $SNAP_DATA or $SNAP_COMMON as oldname.
+		// This is done so that snaps cannot attempt to bypass restrictions
+		// by mounting something outside into their own space.
```

Comment:
Note that apparmor will resolve symlinks, so this is more about usability than confinement bypass.

Reply:
Thanks for confirming that.